internal/controllers/crds/dynamic_controller.go

```diff
@@ -0,0 +1,134 @@
+package crds
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/samber/lo"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/utils"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,verbs=list;watch
+
+type Controller interface {
+	SetupWithManager(mgr ctrl.Manager) error
+}
+
+// DynamicController ensures that RequiredCRDs are installed in the cluster and only then sets up all of its Controllers
+// that depends on them.
+// In case the CRDs are not installed at start-up time, DynamicController will set up a watch for CustomResourceDefinition
+// and will dynamically set up its Controllers once it detects that all RequiredCRDs are already in place.
+type DynamicController struct {
+	Log              logr.Logger
+	Manager          ctrl.Manager
+	CacheSyncTimeout time.Duration
+	Controllers      []Controller
+	RequiredCRDs     []schema.GroupVersionResource
+
+	startControllersOnce sync.Once
+}
+
+func (r *DynamicController) SetupWithManager(mgr ctrl.Manager) error {
+	if r.allRequiredCRDsInstalled() {
```

Comment:
What if someone deletes a CRD afterwards? We'd still be running the controller which is not bad per se - it's ultimately what we do now - but I wonder what would be the cost and whether it would make sense to add functionality to  disable the controllers of CRDs that get deleted.

Reply:
Fair enough. It might not be worth the effort given that it's an unlikely scenario and the benefits might be negligible.